### DISCUSSION TITLE
Is it possible to nest `arg()`?

### DISCUSSION THREAD

**Discussion by ramiroaraujo** [johnlindquist/kit#1157] [Q&A]
For example, I give you a list of options, maybe the last one is "other", and I'd like to require input for the "other" option. I initially thought it to be straightforward but I'm seeing some issues. Maybe I'm tired or maybe it's not possible?

**Comment by johnlindquist** [comment]
@ramiroaraujo 

I'm not sure I understand the question... Were you thinking of something like this?


```js
let result = await arg("Select", ["one", "two", "other"])

if(result ===  "other"){
  result = await arg("More options", ["three", "four"])
}
```

**Reply by ramiroaraujo** [reply]
hmm, I might be over-complicating stuff, but basically I'm porting a string manipulation that I have in Alfred, similar to piping strings in the terminal, so you have a list of operations, each with a transform function, all `(text: string) => string`, and you can perform multiple operations on the same string. UI-wise it keeps prompting for operations until you finish for example with a shortcut.
One of the operations is a regexp, and I want to prompt for input.

Kind of like this:
![image](https://user-images.githubusercontent.com/46276/223237406-70ab8fd3-d462-4384-8dc2-3dfa24205d64.png)

In the RegExp example I'm prompting for the input to capture, and it's not working correctly. Is it ok to do it on the `onSubmit`?

**Reply by johnlindquist** [reply]
Does something like this help? I just kind of winged it, I didn't want to type out the screenshot 😅

```js
// Name: Process Text

import "@johnlindquist/kit"

let previewText = text =>
  `<div class="p-4 whitespace-pre-wrap break-words">${text}</div>`

let matcher = (text, exp) =>
  text.match(new RegExp(exp, "gm"))

let operations = {
  Trim: text => text.trim(),
  Uppercase: text => text.toUpperCase(),
  Lowercase: text => text.toLowerCase(),
  RegExp: async text => {
    let expression = await arg(
      {
        hint: "Type some regex",
      },
      input => {
        let matches = matcher(text, input)
        if (Array.isArray(matches))
          return setPanel(
            previewText(
              matches.filter(m => m.trim()).join("\n")
            )
          )

        return [
          {
            name: "No matches",
            disableSubmit: true,
          },
        ]
      }
    )

    return matcher(text, expression)
      .filter(m => m.trim())
      .join("\n")
  },
}

let text = await editor({
  value: `
    This is just 
    
    
        some text
    
    to experiment with
    
    `.trim(),
  shortcuts: [
    {
      name: "Process Text",
      onPress: submit,
      bar: "right",
      key: "cmd+s",
    },
  ],
})

let flags = {
  done: {
    name: "Done",
    shortcut: "cmd+enter",
  },
}

while (true) {
  let operation = await arg(
    {
      placeholder: "Operate",
      preview: previewText(text),
      flags,
    },
    Object.keys(operations)
  )

  if (flag.done) break

  text = await operations[operation](text)
}

await setSelectedText(text)
```

Re: onSubmit

`onSubmit` can't "prevent" submit from happening. It's just a last chance to grab the current input/state from the app in the handler if needed.

**Reply by ramiroaraujo** [reply]
@johnlindquist Thanks! I'll check it out. I'm just learning the API, I though if `async`ed it would wait until resolved

**Reply by ramiroaraujo** [reply]
Hah, I see you liked it and improved it quite a bit. I love seeing extra tools like the text input playground to see results in real-time :)
